File: src/Cdn/Cache/Adapter/Balancer.php
<?php

namespace Utopia\Cdn\Cache\Adapter;

use Utopia\Balancer\Balancer as OptionBalancer;
use Utopia\Cdn\Cache\Adapter;
use Utopia\Cdn\Domain;
use Utopia\Cdn\Exception\Configuration;
use Utopia\Cdn\Exception\Purge;
use Utopia\Cdn\Exception\UnsupportedOperation;
use Utopia\Cdn\Extend\CdnOption;

/**
 * Purges through every option a balancer's filters leave standing.
 *
 * The balancer holds the full set of configured providers and the filters that
 * narrow it down, which keeps provider selection declarative: a caller states
 * what it wants purged and which options qualify, never which API to call. A
 * domain cached by more than one provider is evicted from all of them, because
 * leaving one provider holding a stale response is the same as not purging.
 *
 * Providers are attempted independently. One failing does not stop the rest,
 * and the collected failures are raised together once every option has been
 * tried, so a single provider outage cannot silently skip the others.
 */
class Balancer implements Adapter
{
    public function __construct(private readonly OptionBalancer $balancer) {}

    public function purgePaths(string $domain, array $paths): void
    {
        $domain = Domain::validate($domain);
        $paths = Domain::validatePaths($paths);

        if ($paths === []) {
            return;
        }

        $this->each('path purging', static function (Adapter $adapter) use ($domain, $paths): void {
            $adapter->purgePaths($domain, $paths);
        });
    }

    public function purgeDomain(string $domain): void
    {
        $domain = Domain::validate($domain);

        $this->each('domain purging', static function (Adapter $adapter) use ($domain): void {
            $adapter->purgeDomain($domain);
        });
    }

    public function purgeKeys(array $keys): void
    {
        if ($keys === []) {
            return;
        }

        $this->each('cache key purging', static function (Adapter $adapter) use ($keys): void {
            $adapter->purgeKeys($keys);
        });
    }

    /**
     * Purges every zone behind a matching option, which is as wide as a purge gets here: each
     * provider drops everything it holds, for every domain, not only the ones these options front.
     */
    public function purgeZone(): void
    {
        $this->each('zone purging', static function (Adapter $adapter): void {
            $adapter->purgeZone();
        });
    }

    /**
     * @param callable(Adapter): void $purge
     */
    private function each(string $operation, callable $purge): void
    {
        $options = $this->balancer->getFilteredOptions();

        if ($options === []) {
            throw new Configuration('No cache options matched the balancer filters.');
        }

        /** @var array<int, \Throwable> $errors */
        $errors = [];
        /** @var array<int, string> $failed */
        $failed = [];
        $purged = false;

        foreach ($options as $option) {
            // A balancer accepts any option, so what it is holding is checked here.
            if (!$option instanceof CdnOption) {
                throw new Configuration('Cache options must be instances of ' . CdnOption::class . '.');
            }

            try {
                $purge($option->getAdapter());
                $purged = true;
            } catch (UnsupportedOperation) {
                // An option that cannot serve this operation is not a failure;
                // the remaining options still have to be purged.
                continue;
            } catch (\Throwable $error) {
                $errors[] = $error;
                $failed[] = $option->getProvider();
            }
        }

        if ($errors !== []) {
            throw new Purge('Cache ' . $operation . ' failed for ' . implode(', ', array_unique($failed)) . '.', $errors);
        }

        if (!$purged) {
            throw new UnsupportedOperation('Cache ' . $operation . ' is not supported by any matching option.');
        }
    }
}

File: 3.x.x/crates-utopia/cdn/src/cache/adapter/balancer.rs
//! PHP `Utopia\Cdn\Cache\Adapter\Balancer`.

use super::Adapter;
use crate::extend::{OptionBalancer, OptionKind};
use crate::{CdnError, Configuration, Domain, Purge, UnsupportedOperation};

/// Purges through every option a balancer's filters leave standing.
pub struct Balancer {
    balancer: OptionBalancer,
}

impl std::fmt::Debug for Balancer {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Balancer").finish_non_exhaustive()
    }
}

impl Balancer {
    #[must_use]
    pub fn new(balancer: OptionBalancer) -> Self {
        Self { balancer }
    }

    fn each<F>(&self, operation: &str, mut purge: F) -> Result<(), CdnError>
    where
        F: FnMut(&dyn Adapter) -> Result<(), CdnError>,
    {
        let options = self.balancer.filtered();
        if options.is_empty() {
            return Err(
                Configuration("No cache options matched the balancer filters.".into()).into(),
            );
        }
        let mut errors = Vec::new();
        let mut failed = Vec::new();
        let mut purged = false;
        for option in &options {
            let OptionKind::Cdn(cdn) = option else {
                return Err(Configuration(
                    "Cache options must be instances of Utopia\\Cdn\\Extend\\CdnOption.".into(),
                )
                .into());
            };
            let adapter = cdn.get_adapter()?;
            match purge(adapter.as_ref()) {
                Ok(()) => purged = true,
                Err(CdnError::UnsupportedOperation(_)) => {}
                Err(err) => {
                    failed.push(cdn.get_provider()?.to_owned());
                    errors.push(err);
                }
            }
        }
        if !errors.is_empty() {
            // PHP `array_unique` keeps first-seen order.
            let mut unique = Vec::new();
            for name in &failed {
                if !unique.contains(name) {
                    unique.push(name.clone());
                }
            }
            return Err(Purge::new(
                format!("Cache {operation} failed for {}.", unique.join(", ")),
                errors,
            )
            .into());
        }
        if !purged {
            return Err(UnsupportedOperation(format!(
                "Cache {operation} is not supported by any matching option."
            ))
            .into());
        }
        Ok(())
    }
}

impl Adapter for Balancer {
    fn purge_paths(&self, domain: &str, paths: &[String]) -> Result<(), CdnError> {
        let domain = Domain::validate(domain)?;
        let paths = Domain::validate_paths(paths)?;
        if paths.is_empty() {
            return Ok(());
        }
        self.each("path purging", |adapter| {
            adapter.purge_paths(&domain, &paths)
        })
    }

    fn purge_domain(&self, domain: &str) -> Result<(), CdnError> {
        let domain = Domain::validate(domain)?;
        self.each("domain purging", |adapter| adapter.purge_domain(&domain))
    }

    fn purge_keys(&self, keys: &[String]) -> Result<(), CdnError> {
        if keys.is_empty() {
            return Ok(());
        }
        self.each("cache key purging", |adapter| adapter.purge_keys(keys))
    }

    fn purge_zone(&self) -> Result<(), CdnError> {
        self.each("zone purging", |adapter| adapter.purge_zone())
    }
}
